### PR TITLE
[public-api] improve request exception logging

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

During rollout we see some functions are failing with Unknown reason, but they are not logged on server side and there is no information to correlate them from a client side.  This PR ensures that all exceptional situations are logged on server side, plus on client side we provide request id and method to look up the entire chain in logs.

It also fixes a missing conversion for invalid argument which is reported as Unknown reason right now.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 895242f</samp>

This pull request improves the error handling, tracing, and documentation of the public API communication between the dashboard and the server components. It uses the `handleError` and `PublicAPIConverter` functions to report errors consistently and with context information. It also adds a README file to the `components/public-api/typescript-common` package and updates the test case for the error message change.

</details>

## Related Issue(s)

Fixes #

## How to test

- Check that each gRPC request has `x-request-id` response header (for any status code)
Screenshot 2023-12-01 at 11 34 05

- Trigger invalid argument from browser console. Check that it is reported as 400 now, not unknown error. Check that server logs don't log it as `public api exception`.
```
fetch("https://ak-request9e413561ca.preview.gitpod-dev.com/public-api/gitpod.v1.OrganizationService/GetOrganizationSettings", {
  "headers": {
    "connect-protocol-version": "1",
    "content-type": "application/json",
  },
  "referrer": "https://ak-request9e413561ca.preview.gitpod-dev.com/settings",
  "referrerPolicy": "no-referrer-when-downgrade",
  "body": "{\"organizationId\":\"xyz\"}",
  "method": "POST",
  "mode": "cors",
  "credentials": "include"
});
```
- Go to workspaces page, check that you see WatchWorkspaceStatus or LotsOfReplies stream, and then delete server pod to trigger its failure. Check that `reportError` has info about `requestId` as well as `requestMethod`.
Screenshot 2023-12-01 at 12 49 31

#### Preview status

https://ak-request9e413561ca.preview.gitpod-dev.com/workspaces

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/dashboard/src/service/public-api.ts
Comment (geropl) [inline components/dashboard/src/service/public-api.ts:200]: :grimacing:

File: components/server/src/api/server.ts
Comment (geropl) [inline components/server/src/api/server.ts:162]: Hm. It feels like we should separate these two IDs: The one we use internally should be :100: under our control. And if we receive one from the client that's cool, we can add it as separate field.

File: components/public-api/typescript-common/README.md
Comment (geropl) [inline components/public-api/typescript-common/README.md:7]: :orange_heart:

**Comment by geropl** [comment]
Starting preview...

**Review by geropl — APPROVED** [review]
Code LGTM (sans this comment), tested and works. :checkered_flag: 

:heavy_check_mark: to unblock (re-ping if you want another review :+1:)

**Comment by akosyakov** [comment]
@geropl it overcomplicated it, but using proper APIs even with stream response headers are enough. I pushed an update. Fine to unhold?

If server stream fails after succesful connection requestId is present:
Screenshot 2023-12-01 at 17 17 24

If server stream is not able to connect at all requestId is missing:
Screenshot 2023-12-01 at 17 17 29

**Pull request merged**
